### PR TITLE
Read compressed FITS files from file handles

### LINKED ISSUE #5948: fits.open() doesn't support compression on already open file-handles
When opening a FITS file on disk by providing the path string:
```python
fits.open("my_file.fits.gz")
```
compression in the files is handled automatically. This is taken care of by the `_File` class, which appears to check either for the '.gz' extension or for the right "magic" number at the start of the file.

However, this logic is not present when using a file handle, e.g.:
```python
file_object = open("tmp.fits.gz")
f = fits.open(file_object)  # raises IOError
```

Here is a MWE:

```pthon
from astropy.io import fits
import numpy as np

# Create gzipped fits file:
fits.PrimaryHDU(np.random.random((20, 20))).writeto("tmp.fits.gz")

# Open using the usual way (no errors):
f = fits.open("tmp.fits.gz")

# Open using a filehandle:
file_object = open("tmp.fits.gz")
f = fits.open(file_object)  # raises IOError
```

See http://stackoverflow.com/questions/43220542/how-to-open-a-remote-fts-gz-file-with-astropy-io-fits-open for a particular use-case.

### PR DESCRIPTION
**Pull request by drdavella**
This fixes #5948. It is now possible to open compressed FITS files using file handles:
```python
from astropy.io import fits

with open('example.fits.gz', 'rb') as fh:
    # Automatically detects that the file handle corresponds to a gzip file
    with fits.open(fh) as fits_handle:
        # do stuff
```

This was already possible when calling `fits.open` with a string corresponding to a path or with a compression file object (e.g. `gzip.GzipFile`). This PR also adds a number of unit tests to verify the new functionality and to improve coverage of existing functionality.

### CODE REVIEW

**Comment by drdavella** [comment]
The test failures seem to be due to an API difference in `bz2` between Python 2.7 and 3.0. I can probably make it work for 2.7, but how should we proceed given that Python 2.7 support is going away in astropy 3.0?

**Comment by bsipocz** [comment]
I suggest simply to skip the tests if it's python 2.7, and we will clean it up together with the rest during the 2.7 support removal.

File: astropy/io/fits/tests/test_core.py [hunk 1]
```diff
@@ -8,6 +8,7 @@
 import os
 import warnings
 import zipfile
+import six
```
Comment (bsipocz) [inline astropy/io/fits/tests/test_core.py]: please use the bundled version of six

**Comment by pllim** [comment]
Is this still a work in progress or ready for review?

**Comment by drdavella** [comment]
It's ready for review.

File: astropy/io/fits/file.py [hunk 2]
```diff
@@ -426,13 +430,33 @@ def _open_fileobj(self, fileobj, mode, overwrite):
...
-            # append mode the file pointer is at the end of the file
+        # Attempt to determine if the file represented by the open file object
+        # is compressed
+        try:
+            magic = self._file.read(4)
             self._file.seek(0)
+        except (IOError,OSError):
+            return
+
+        if magic.startswith(GZIP_MAGIC):
```
Comment (MSeifert04) [inline astropy/io/fits/file.py]: That part (starting here) is copied from the other method `_open_filename`, correct? 

Maybe it would be better to refactor that into a seperate (private method) if it's copied. That makes future maintenance easier.

File: astropy/io/fits/file.py [hunk 3]
```diff
@@ -109,15 +109,19 @@ def __init__(self, fileobj=None, mode=None, memmap=None, overwrite=False,
...
-                # If the mode is unsupported just leave it as None; we'll
-                # catch this case below
-                mode = FILE_MODES.get(fmode)
+                mode = fileobj_mode(fileobj)
             else:
                 mode = 'readonly'  # The default
 
         if mode not in IO_FITS_MODES:
-            raise ValueError("Mode '{}' not recognized".format(mode))
+            if mode in ['r', 'w']:
```
Comment (MSeifert04) [inline astropy/io/fits/file.py]: What about `append` mode? I'm a bit hazy on the mode details but wouldn't `append` be `'a'` mode?

File: astropy/io/fits/tests/test_core.py [hunk 4]
```diff
@@ -593,15 +594,48 @@ def test_open_nonexistent(self):
...
+        # Make sure we can open a FITS file from an open file handle
+        with open(self.data('test0.fits'), 'rb') as handle:
+            with fits.open(handle) as fitsfile:
+                pass
+
+        with open(self.temp('temp.fits'), 'wb') as handle:
+            with fits.open(handle) as fitsfile:
+                pass
+
+        # Opening without explicitly specifying binary mode should fail
```
Comment (MSeifert04) [inline astropy/io/fits/tests/test_core.py]: I'm not sure if `binary` should be always required. If I remember correctly text-mode and binary-mode are only different for windows-machines.

But I'm not so familiar with the modes anymore so it would be okay with me to keep that. :)

Comment (drdavella) [inline astropy/io/fits/tests/test_core.py]: There is a difference: see [this](https://docs.python.org/3.6/tutorial/inputoutput.html#reading-and-writing-files) from the Python documentation for more information.

I'll admit I'm not familiar with all of the use cases, but if the file to be processed has any chance of containing non-decodable binary data then it is important to open with `'b'` or else later operations will fail.

File: astropy/io/fits/file.py (hunk 2 above)
Comment (drdavella) [inline astropy/io/fits/file.py]: It's not identical but I'll see if there's a way to share more code between the two cases.

File: astropy/io/fits/file.py (hunk 3 above)
Comment (drdavella) [inline astropy/io/fits/file.py]: That occurred to me too but I forgot to make the update.

File: astropy/io/fits/file.py (hunk 2 above)
Comment (drdavella) [inline astropy/io/fits/file.py]: Done. Hopefully it's a bit cleaner now.

File: astropy/io/fits/file.py [hunk 5]
```diff
@@ -400,6 +404,36 @@ def _overwrite_existing(self, overwrite, fileobj, closed):
             else:
                 raise IOError("File {!r} already exists.".format(self.name))
 
+    def _test_for_compression(self, obj_or_name, magic, mode, ext=None):
```
Comment (MSeifert04) [inline astropy/io/fits/file.py]: Maybe `ext` should be `''` by default. On Python3 one can't compare strings to `None`.

Comment (drdavella) [inline astropy/io/fits/file.py]: That's a reasonable change, although I'll admit that I'm not sure I see a technical reason why comparison with `None` in this case is a problem.

**Comment by MSeifert04** [comment]
I totally forgot: Thanks @drdavella for making the PR!

File: astropy/io/fits/file.py (hunk 5 above)
Comment (MSeifert04) [inline astropy/io/fits/file.py]: I was wrong, it doesn't throw an exception. I thought the comparisons `ext == '.gz'` would error out if `ext = None`.

Comment (drdavella) [inline astropy/io/fits/file.py]: I changed it anyway since it clarifies the intent. Maybe you were thinking about how if you actually need to test if an object is `None` you need to use `is` rather than `==`?

Comment (MSeifert04) [inline astropy/io/fits/file.py]: I was thinking about ordering operators in general (which throw exceptions with incompatible types) and forgot that `==` and `!=` do work even if the types aren't orderable.

File: astropy/io/fits/file.py [hunk 6]
```diff
@@ -487,22 +521,7 @@ def _open_filename(self, filename, mode, overwrite):
```
Comment (saimn) [inline astropy/io/fits/file.py]: You could move the magic bytes reading here in the new method (`_test_for_compression`).

File: astropy/io/fits/file.py [hunk 7]
```diff
@@ -487,22 +521,7 @@ def _open_filename(self, filename, mode, overwrite):
...
-        elif ext == '.bz2' or magic.startswith(BZIP2_MAGIC):
-            # Handle bzip2 files
-            if mode in ['update', 'append']:
-                raise IOError("update and append modes are not supported "
-                              "with bzip2 files")
-            # bzip2 only supports 'w' and 'r' modes
-            bzip2_mode = 'w' if mode == 'ostream' else 'r'
-            self._file = bz2.BZ2File(self.name, bzip2_mode)
-        else:
+        if not self._test_for_compression(self.name, magic, mode, ext=ext):
```
Comment (saimn) [inline astropy/io/fits/file.py]: The method name is confusing, it doesn't only test, it also read the file if it is compressed. So maybe something like `_try_read_compressed` ?

File: astropy/io/fits/file.py [hunk 8]
```diff
@@ -400,6 +404,33 @@ def _overwrite_existing(self, overwrite, fileobj, closed):
...
+        elif ext == '.bz2' or magic.startswith(BZIP2_MAGIC):
+            # Handle bzip2 files
+            if mode in ['update', 'append']:
+                raise IOError("update and append modes are not supported "
+                              "with bzip2 files")
+            # bzip2 only supports 'w' and 'r' modes
+            bzip2_mode = 'w' if mode == 'ostream' else 'r'
+            self._file = bz2.BZ2File(obj_or_name, mode=bzip2_mode)
+            self.compression = 'bzip2'
+        return self.compression # Will be 'None' if no compression detected
```
Comment (saimn) [inline astropy/io/fits/file.py]: Return `False` to be sure, or maybe raise an exception (what was the previous behavior ?)
`is_compressed` is not used.

File: astropy/io/fits/file.py [hunk 9]
```diff
@@ -578,7 +597,7 @@ def _open_zipfile(self, fileobj, mode):
 
         if close:
             zfile.close()
-        self.compression = 'zip'
+        self._file.seek(0)
```
Comment (saimn) [inline astropy/io/fits/file.py]: First line is moved in the other function, which looks good, but why adding a seek here ? (it's already done in `_open_filename/filelike`)

File: astropy/io/fits/hdu/hdulist.py [hunk 10]
```diff
@@ -27,7 +27,7 @@
 from ....extern.six.moves import range
 
 
-def fitsopen(name, mode='readonly', memmap=None, save_backup=False,
+def fitsopen(name, mode=None, memmap=None, save_backup=False,
```
Comment (saimn) [inline astropy/io/fits/hdu/hdulist.py]: Why ?

File: astropy/io/fits/tests/test_core.py (hunk 4 above)
Comment (saimn) [inline astropy/io/fits/tests/test_core.py]: Binary mode is required, but it's tricky because it was the default on Python 2.

File: astropy/io/fits/hdu/hdulist.py (hunk 10 above)
Comment (saimn) [inline astropy/io/fits/hdu/hdulist.py]: Ok I think I know, because it skips the fileobj mode detection ? There is something weird here, if FITS mode is `readonly` but you pass a file handle with `w` it is not detected.

File: astropy/io/fits/file.py [hunk 11]
```diff
@@ -109,15 +109,19 @@ def __init__(self, fileobj=None, mode=None, memmap=None, overwrite=False,
...
-                # If the mode is unsupported just leave it as None; we'll
-                # catch this case below
-                mode = FILE_MODES.get(fmode)
+                mode = fileobj_mode(fileobj)
             else:
                 mode = 'readonly'  # The default
 
         if mode not in IO_FITS_MODES:
-            raise ValueError("Mode '{}' not recognized".format(mode))
+            if mode in ['r', 'w', 'a']:
```
Comment (drdavella) [inline astropy/io/fits/file.py]: You are right about the `"*t"`options: they do need to be handled (I had forgotten that was allowed actually). This was really just an attempt to refactor the existing logic to be a little bit less redundant, and personally I think it's better with the more explicit error message.

File: astropy/io/fits/file.py (hunk 9 above)
Comment (drdavella) [inline astropy/io/fits/file.py]: The seek in `_open_filename/filelike` is necessary in all cases in order to reset the position of the file handle after attempting to read the magic value. However, zip files are handled differently than either `bzip` or `gzip` because it is necessary to read the first entry in the zip archive and write it to a new file. You'll notice in the code above that the last operation on the underlying file handle was a write. So in order for the HDU detection to work properly, it is necessary to seek back to the beginning of the new file. I'll add a comment to clarify.

Comment (drdavella) [inline astropy/io/fits/file.py]: You're right though that there is now a redundant seek in `_open_filelike`, which can be removed.

File: astropy/io/fits/file.py [hunk 12]
```diff
@@ -426,13 +457,16 @@ def _open_fileobj(self, fileobj, mode, overwrite):
...
-        else:
-            self._file = gzip.open(self.name, IO_FITS_MODES[mode])
 
-        if fmode == 'ab+':
-            # Return to the beginning of the file--in Python 3 when opening in
-            # append mode the file pointer is at the end of the file
+        # Attempt to determine if the file represented by the open file object
+        # is compressed
+        try:
+            magic = self._file.read(4)
```
Comment (drdavella) [inline astropy/io/fits/file.py]: The seek afterwards is definitely necessary in order to restore the file pointer after attempting to read the magic values. I'm not sure about a seek before reading the first bytes. The only way it would matter is if someone had read the file before passing it to `fits.open`, in which case you could argue that they deserve the error they will get.

I actually think that handling an exception here is a good idea. If there is a problem, it will still cause an error later on because the FITS header will not appear to be valid. But it will be more helpful than receiving a mysterious `IOError` here which could potentially be harder for a user to debug.

File: astropy/io/fits/file.py (hunk 7 above)
Comment (drdavella) [inline astropy/io/fits/file.py]: Yes, good idea.

File: astropy/io/fits/file.py (hunk 6 above)
Comment (drdavella) [inline astropy/io/fits/file.py]: The magic is read differently based on whether it's in `_open_fileobj` or `_open_filename` so I think it's best left as is.

File: astropy/io/fits/file.py (hunk 8 above)
Comment (drdavella) [inline astropy/io/fits/file.py]: Yes, good point.

File: astropy/io/fits/hdu/hdulist.py (hunk 10 above)
Comment (drdavella) [inline astropy/io/fits/hdu/hdulist.py]: Good catch, this is a bit of a mess. I'll work on it and commit a fix.

**Comment by drdavella** [comment]
@saimn I pushed some updates based on your feedback. I think I managed to clean up the way that FITS thinks about file modes, but maybe there's still something I missed.

It looks like I need to rebase this branch on the current master though since you merged #6378. Should I go ahead and do that?

**Comment by saimn** [comment]
@drdavella - Yes, please rebase.

File: astropy/io/fits/file.py [hunk 13]
```diff
@@ -49,6 +50,9 @@
     'wb': 'ostream', 'wb+': 'update',
     'ab': 'ostream', 'ab+': 'append'}
 
+# A match indicates the file was opened in text mode, which is not allowed
+TEXT_RE = re.compile('^[rwa]((t?\+?)|(\+?t?))$')
```
Comment (bsipocz) [inline astropy/io/fits/file.py]: use raw string here to keep python 3.6 happy.

Comment (drdavella) [inline astropy/io/fits/file.py]: I updated with this fix and did a little bit of squashing to clean up.

Comment (pllim) [inline astropy/io/fits/file.py]: There is a conflict with `master` now. In regards to squashing, IMHO it is best to do it as a last step when ready to merge; This is because reviewer sometimes just want to look at the last commit(s) since review, so squashing would cause unnecessary re-review.

Comment (drdavella) [inline astropy/io/fits/file.py]: Do you want me to rebase again?

Comment (pllim) [inline astropy/io/fits/file.py]: Yes, please. 😄

Comment (bsipocz) [inline astropy/io/fits/file.py]: yes please, do rebase whenever there is a conflict.

Comment (drdavella) [inline astropy/io/fits/file.py]: Okay, I just wanted to check since the development workflow says to rebase [only if asked](http://docs.astropy.org/en/stable/development/workflow/development_workflow.html#rebase-but-only-if-asked) :smile:

Comment (bsipocz) [inline astropy/io/fits/file.py]: I've guessed that. Let's say that guide is not aimed at advanced contributors like you, but more for people who are new to git and the basic concepts of PR based development.

**Comment by bsipocz** [comment]
@drdavella - Could you rebase once again?

@saimn - Could you review and approve it if you think it's good to go?

**Comment by drdavella** [comment]
@bsipocz, done.

**Comment by drdavella** [comment]
I'm revisiting this one in the hopes that it will get merged soon. I'm not sure I understand the results from coveralls, though. Is it possible that I just need to rebase again?

**Comment by saimn** [comment]
@drdavella - I will try to review soon, with holidays I did not manage to do this before, so now I need to plunge back into this :wink:

**Comment by drdavella** [comment]
@saimn no worries and no rush! I am mostly just confused about the coveralls report and was wondering if I should rebase to see if it sorts itself out.

**Comment by saimn** [comment]
@drdavella - Yep, a rebase can probably fix that (it seems that Coveralls compare the coverage to the one from the master branch, not from the branch base commit).

**Comment by bsipocz** [comment]
@drdavella - Yes, coveralls is rather dumb, and doesn't take it into account if master changed significantly. I'll try to look at and try out other coverage services, maybe we can find something better.

**Comment by drdavella** [comment]
It looks like this was just a transient test error on travis 😒 . Does someone mind restarting?

File: astropy/io/fits/tests/test_core.py [hunk 14]
```diff
@@ -1068,7 +1155,7 @@ def get_free_space_in_dir(path):
             hdulist = fits.HDUList(hdu)
             filename = self.temp('test.fits')
 
-            with open(filename, mode='wb+') as fileobj:
+            with open(filename, mode='wb') as fileobj:
```
Comment (saimn) [inline astropy/io/fits/tests/test_core.py]: Why this change ?

File: astropy/io/fits/file.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (saimn) [reply astropy/io/fits/file.py]: > The seek afterwards is definitely necessary in order to restore the file pointer after attempting to read the magic values

Right, but I don't see how the previous seek for `fmode == 'ab+'` is now done ?

> I'm not sure about a seek before reading the first bytes

Indeed, maybe not useful.

**Comment by drdavella** [comment]
> this code is complicated, with many possible paths, and it makes the diff very difficult to follow

@saimn, I'm experiencing the same thing. Even though I worked on this, it has been a while since I looked at it and so I have to refresh my memory about why certain decisions were made.

File: astropy/io/fits/tests/test_core.py (hunk 14 above)
Comment (drdavella) [reply astropy/io/fits/tests/test_core.py]: This is a bit complicated: the short answer is that the previous use of `'wb+'` was incorrect, but the FITS file mode checking logic was not robust enough to detect the error.

The `writeto` method is used to create a new FITS file from scratch. This means the file is opened as an `'ostream'` in FITS terminology. To FITS, `'wb'` means `'ostream'`, whereas `'wb+'` means `'update'`. So the mode that the file was opened in was not in agreement with the way that FITS wanted to use it. Previously this error was allowed to pass unchecked, but with the more robust mode checking, it actually causes an error.

**Comment by saimn** [comment]
> @saimn, I'm experiencing the same thing.

Yeah, sorry about the long delay before coming back to this !

File: astropy/io/fits/file.py [hunk 15]
```diff
@@ -152,7 +166,7 @@ def __init__(self, fileobj=None, mode=None, memmap=None, overwrite=False,
         self.writeonly = False
 
         # Initialize the internal self._file object
-        if _is_random_access_file_backed(fileobj):
+        if isfile(fileobj):
```
[the remark this answers is not in the document]
Comment (drdavella) [reply astropy/io/fits/file.py]: If a `GzipFile` is given, this logic will be skipped. If `fileobj` is a `GzipFile`, it will be handled on line 178.

File: astropy/io/fits/file.py [hunk 16]
```diff
@@ -76,6 +80,17 @@
...
 
+def _normalize_fits_mode(mode):
+    if mode is not None and mode not in IO_FITS_MODES:
+        if TEXT_RE.match(mode):
+            raise ValueError(
+                "Text mode '{}' not supported: "
+                "files must be opened in binary mode".format(mode))
+        new_mode = FILE_MODES.get(mode)
+        if new_mode not in IO_FITS_MODES:
+            raise ValueError("Mode '{}' not recognized".format(mode))
```
[the remark this answers is not in the document]
Comment (drdavella) [reply astropy/io/fits/file.py]: In order to properly handle all types of file objects, the default cannot actually be `readonly` in all cases. If a path or URL is passed, then the effective default is still `readonly`, so nothing has changed. However, if a file object is passed, we want the default to match the mode of the given file object. For example, if I open a file with `'wb'` and pass the handle to `fitsopen`, I don't want my default to be `readonly`, I want it to be `ostream`.

With respect to the file mode check in `_open_fileobj`, that may be an oversight on my part but I need to take a closer look. It's likely that that check is now redundant and should be removed. In any case, that check does not happen early enough to catch erroneous cases where the mode of the given `fileobj` does not match the requested mode.

I attempted to thoroughly test all of these combinations and interactions in the new unit tests (which were added since the last time this was reviewed). However, it's possible that I missed some edge cases, in which case I'd be happy to add more tests.

File: astropy/io/fits/file.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (drdavella) [reply astropy/io/fits/file.py]: I might argue that if `fileobj` is a file handle that was opened with `mode='ab'` or `mode='ab+'`, that is not a valid use case and it should result in an error in `__init__`. It just doesn't make any sense to do that. That's not currently being checked, though.

File: astropy/io/fits/file.py [hunk 17]
```diff
@@ -408,6 +422,32 @@ def _overwrite_existing(self, overwrite, fileobj, closed):
...
+                raise IOError("update and append modes are not supported "
+                              "with bzip2 files")
+            # bzip2 only supports 'w' and 'r' modes
+            bzip2_mode = 'w' if mode == 'ostream' else 'r'
+            self._file = bz2.BZ2File(obj_or_name, mode=bzip2_mode)
+            self.compression = 'bzip2'
+        return self.compression is not None
+
     def _open_fileobj(self, fileobj, mode, overwrite):
         """Open a FITS file from a file object or a GzipFile object."""
```
[the remark this answers is not in the document]
Comment (drdavella) [reply astropy/io/fits/file.py]: Good catch, will update.

File: astropy/io/fits/file.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (saimn) [reply astropy/io/fits/file.py]: > However, if a file object is passed, we want the default to match the mode of the given file object. For example, if I open a file with 'wb' and pass the handle to fitsopen, I don't want my default to be readonly, I want it to be ostream.

Currently the behavior is that the file object mode must match the "fits mode", not the opposite. If the file mode does not match the fits mode (with the test in `_open_fileobj`) then an error is raised. 
And the fits mode is determined from the file mode only if fits mode is `None`.

So with your PR and comment cited above I'm not sure if you try to invert this logic, and why this would be needed for the original purpose, reading compressed file objects. Maybe it's just that the check in `_open_fileobj` was missing a few cases, or that check should be done earlier ?

File: astropy/io/fits/tests/test_core.py (hunk 14 above)
Comment (saimn) [reply astropy/io/fits/tests/test_core.py]: Got it, thanks.

Comment (drdavella) [reply astropy/io/fits/tests/test_core.py]: Any chance this clarifies the question you asked above? I'm trying to write an explanation but keep failing at words 😛

File: astropy/io/fits/file.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (drdavella) [reply astropy/io/fits/file.py]: Here's the basic problem and an outline of the process I had to go through to make the compressed files work.: 
* if we use 'readonly' as the default, then we don't know if the user explicitly provided 'readonly' as the FITS mode, or whether no FITS mode was provided and we needed to choose a default. 
* It turns out that 'readonly' is not a valid default in all cases: if a file handle is provided, the default depends on the mode that was used to open that file handle.
* Once I fixed this, I realized that agreement between FITS mode and the underlying file mode was not being checked properly in all cases
* Once I made the check work properly, the check in `_open_fileobj` became redundant

If the FITS default mode handling and mode checking had been working properly in the first place, then this PR would have been **much** simpler 😅

**Comment by drdavella** [comment]
@saimn sorry for the poke but just wondering if you think this is ready for merge yet. I hope I've convinced you that all the changes here are good and necessary but let me know if you have other questions.

File: astropy/io/fits/file.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (saimn) [reply astropy/io/fits/file.py]: Thanks for details @drdavella ! I will also try to explain better what bothers me with this change :wink: 

> Here's the basic problem and an outline of the process I had to go through to make the compressed files work.:
>   - If we use 'readonly' as the default, then we don't know if the user explicitly provided 'readonly' as the FITS mode, or whether no FITS mode was provided and we needed to choose a default. This matters when we pass a compressed file handle.

There is no need "to choose a default", if the file mode is not compatible with the FITS mode (*) an exception must be raised. That's the current behavior (for file handles) and I don't see why it should be different with a *compressed* file handle ?

(*) - just to be clear on terminology, I mean the `mode` argument for `fitsopen` ('readonly' etc.)

>   - It turns out that 'readonly' is not a valid default in all cases: if a file handle is provided, the default depends on the mode that was used to open that file handle.

Nope, the file mode must be in agreement with the FITS mode. That's the root of the issue for me: with your changes, the file handle mode becomes more important than the FITS mode. And the FITS mode becomes useless (for file handles) as you set it to None by default and guess it from the file mode.

Also, I'm not saying that this way to proceed is wrong, but it may have unexpected consequences on existing codes (even if the probability is weak, when dealing with file modes you can have surprises), and this change seems unnecessary with respect to the purpose of the PR. 

Maybe I'm wrong here, but from what I understand the meaning of the `mode` argument for `fitsopen` is different now. For file handles it becomes a way to check the file handle mode. 

>   - Once I fixed this, I realized that agreement between FITS mode and the underlying file mode was not being checked properly in all cases
>   - Once I made the check work properly, the check in _open_fileobj became redundant

> If the FITS default mode handling and mode checking had been working properly in the first place, then this PR would have been much simpler :sweat_smile:

Sure, it's certainly possible that file mode agreement was not checked strictly enough, so fixing this is good!

[the remark this answers is not in the document]
Comment (drdavella) [reply astropy/io/fits/file.py]: Okay, I think I understand your objection now. Will you accept changing the `mode` argument for `fitsopen` back to `'readonly'` as a solution? 

I will also need to remove one of the test cases in `astropy/io/fits/tests/test_core.py:test_open_file_handle` since it will no longer pass. However, according to your argument, the behavior it was testing was not expected to begin with. 

In particular, this tested whether it is possible to open a FITS file from a file handle that had been opened with `'wb'`. It sounds like you are saying that this should **not** be possible unless the `mode='ostream'` argument is explicitly provided to `fitsopen`. Is that correct?

[the remark this answers is not in the document]
Comment (saimn) [reply astropy/io/fits/file.py]: Yes exactly!

**Comment by bsipocz** [comment]
I've cancelled OSX, everything else was passing on travis.

**Comment by drdavella** [comment]
@saimn hopefully you can approve this change now. I apologize for making this so difficult but i had an incorrect assumption about how the defaults were supposed to work. I think that it is correct now, though.

File: astropy/io/fits/file.py [hunk 18]
```diff
@@ -109,17 +124,16 @@ def __init__(self, fileobj=None, mode=None, memmap=None, overwrite=False,
         # Holds mmap instance for files that use mmap
         self._mmap = None
 
+        mode = _normalize_fits_mode(mode)
```
[the remark this answers is not in the document]
Comment (drdavella) [reply astropy/io/fits/file.py]: I'm not sure this is entirely true: we do allow people to use things like 'rb' or 'wb' for the FITS mode itself right? So we need to be able to convert that to a 'normalized' FITS mode that is understood internally. Maybe this is just another wrong assumption I have about use cases but I thought that this behavior was supported even before this PR.

[the remark this answers is not in the document]
Comment (saimn) [reply astropy/io/fits/file.py]: No, the `mode` parameter for `fits.open` should be only in the keys of `IO_FITS_MODES` (`readonly` etc.). That was checked by the `if mode not in IO_FITS_MODES:` block previously (lines 121-122).

[the remark this answers is not in the document]
Comment (drdavella) [reply astropy/io/fits/file.py]: Okay. In principle I don't see why this shouldn't be possible since there is already a well-defined mapping between conventional file modes and FITS modes, but I will change this now in order to get this PR approved.

[the remark this answers is not in the document]
Comment (saimn) [reply astropy/io/fits/file.py]: The mapping is not 1 to 1, `rb` can be used for readonly, copyonwrite, and denywrite. I think that the main reason for these specific names, and imo it would be confusing to support the usual file modes in addition. We can still discuss that in another issue if you want, but here it's better to focus on reading compressed files :wink:

**Comment by saimn** [comment]
Tests pass (except the osx build) so let's move forward and merge this. Thanks for your patience @drdavella !

**Comment by drdavella** [comment]
@saimn thank you for your patience too!
